### PR TITLE
use separate page tables to manage different type of memory

### PR DESCRIPTION
**Pull request by ryuxin**
### Summary of this PR

support multiple page tables to manage different type of memory.
current implementation uses a separate page tables to manage local memory.
modify cos_meminfo struct to include a cap to page table which provides memory.
some explanation to the page table capability.
meta (struct cos_compinfo \*memsrc) maintains the manager component's information, the pgtbl_cap in meta is the capability to the manager's own page table.
The pgtbl_cap in **local_mi** is the capability to a page table which tracks memory resource, currently it tracks all local memory.
**ci** maintains the information of a random client component. The pgtbl_cap in **ci** a capability to client's page table, and that capability is **within manager's (meta) capability table.**
The same logic applies to the captbl_cap

### Code Quality

As part of this pull request, I've considered the following:

Style:

- [x] Comments adhere to the Style Guide (SG)
- [x] Spacing adhere's to the SG 
- [x] Naming adhere's to the SG
- [x] All other aspects of the SG are adhered to, or exceptions are justified in this pull request

Code Craftsmanship:

- [x] I've made an attempt to remove all redundant code
- [x] I've considered ways in which my changes might impact existing code, and cleaned it up
- [x] I've formatted the code in an effort to make it easier to read (proper error handling, function use, etc...)
- [x] I've commented appropriately where code is tricky
- [x] I agree that there is no "throw-away" code, and that code in this PR is of high quality

### CODE REVIEW

File: src/components/implementation/tests/micro_booter/micro_booter.c [hunk 1]
```diff
@@ -568,7 +568,7 @@ term_fn(void *d)
 void
 cos_init(void)
 {
-	cos_meminfo_init(&booter_info.mi, BOOT_MEM_KM_BASE, COS_MEM_KERN_PA_SZ);
+	cos_meminfo_init(&booter_info.local_mi, BOOT_MEM_KM_BASE, COS_MEM_KERN_PA_SZ, BOOT_CAPTBL_SELF_LOCAL_PT);
```
Comment (gparmer) [inline src/components/implementation/tests/micro_booter/micro_booter.c]: I thought we discussed making local default, and "global" the one that causes changes of variable names.  Local is the one that is typical.  Global is the one that is aberrant.  Why would we change all of the code to make local the one that (from a variable name-space perspective) is special cased?

File: src/components/include/cos_kernel_api.h [hunk 2]
```diff
@@ -39,16 +40,16 @@ struct cos_compinfo {
 	vaddr_t vas_frontier, vasrange_frontier;
 	/* the source of memory */
 	struct cos_compinfo *memsrc; /* might be self-referential */
-	struct cos_meminfo mi;	     /* only populated for the component with real memory */
+	struct cos_meminfo local_mi;	     /* only populated for the component with real memory */
```
Comment (gparmer) [inline src/components/include/cos_kernel_api.h]: Or do you mean something else by "local" here?

File: src/kernel/include/shared/cos_types.h [hunk 3]
```diff
@@ -235,16 +236,17 @@ static inline unsigned long captbl_idsize(cap_t c)
...
-	BOOT_CAPTBL_KM_PTE     = 16,
-
-	BOOT_CAPTBL_COMP0_CT           = 18,
-	BOOT_CAPTBL_COMP0_PT           = 20,
+	BOOT_CAPTBL_SRET          = 0,
+	BOOT_CAPTBL_SELF_CT       = 4,
+	BOOT_CAPTBL_SELF_PT       = 6,
+	BOOT_CAPTBL_SELF_COMP     = 8,
+	BOOT_CAPTBL_BOOTVM_PTE    = 12,
+	BOOT_CAPTBL_SELF_LOCAL_PT = 14,
```
Comment (gparmer) [inline src/kernel/include/shared/cos_types.h]: shouldn't SELF_PT be local, and SELF_GLOBAL_PT be global?  Didn't we already talk about this?

If some of the comments don't make sense, ask.  I don't like making the same comments multiple times.

File: src/components/lib/cos_kernel_api.c [hunk 4]
```diff
@@ -421,7 +422,7 @@ cos_captbl_alloc(struct cos_compinfo *ci)
 	assert(ci);
 
 	if (__alloc_mem_cap(ci, CAP_CAPTBL, &kmem, &cap)) return 0;
-	if (call_cap_op(ci->captbl_cap, CAPTBL_OP_CAPTBLACTIVATE, cap, ci->pgtbl_cap, kmem, 0)) BUG();
+	if (call_cap_op(ci->captbl_cap, CAPTBL_OP_CAPTBLACTIVATE, cap, __compinfo_metacap(ci)->local_mi.pgtbl_cap, kmem, 0)) BUG();
```
Comment (gparmer) [inline src/components/lib/cos_kernel_api.c]: I'm not sure how this worked before as well.  Perhaps @phani didn't use this with a meta.

File: src/platform/i386/boot_comp.c [hunk 5]
```diff
@@ -15,14 +15,17 @@ extern u8_t *boot_comp_pgd;
 int boot_nptes(unsigned int sz) { return round_up_to_pow2(sz, PGD_RANGE)/PGD_RANGE; }
 
 int
-boot_pgtbl_mappings_add(struct captbl *ct, pgtbl_t pgtbl, capid_t ptecap, const char *label,
+boot_pgtbl_mappings_add(struct captbl *ct, capid_t pgdcap, capid_t ptecap, const char *label,
```
Comment (gparmer) [inline src/platform/i386/boot_comp.c]: why remove the type?

File: src/platform/i386/boot_comp.c [hunk 6]
```diff
@@ -15,14 +15,17 @@ extern u8_t *boot_comp_pgd;
...
 			void *kern_vaddr, unsigned long user_vaddr, unsigned int range, int uvm)
 {
 	int ret;
 	u8_t *ptes;
 	unsigned int nptes = 0, i;
-	struct cap_pgtbl *pte_cap;
+	struct cap_pgtbl *pte_cap, *pgd_cap;
+	pgtbl_t pgtbl;
 
+	pgd_cap = (struct cap_pgtbl*)captbl_lkup(ct, pgdcap);
```
Comment (gparmer) [inline src/platform/i386/boot_comp.c]: `if (CAP_TYPECHK(...)) assert(0);`

have to make sure it is of the correct type.

File: src/platform/i386/boot_comp.c [hunk 7]
```diff
@@ -112,7 +115,7 @@ kern_boot_comp(void)
         struct captbl *ct;
         unsigned int i;
 	u8_t *boot_comp_captbl;
-	pgtbl_t pgtbl = (pgtbl_t)chal_va2pa(&boot_comp_pgd);
+	pgtbl_t pgtbl = (pgtbl_t)chal_va2pa(&boot_comp_pgd), local_pgd;
```
Comment (gparmer) [inline src/platform/i386/boot_comp.c]: I don't think this variable is being used at all.  Please always do a `git diff` before committing.  Clean up things like this.

File: src/platform/i386/boot_comp.c [hunk 8]
```diff
@@ -144,7 +147,7 @@ kern_boot_comp(void)
 
 	printk("\tCapability table and page-table created.\n");
 
-	ret = boot_pgtbl_mappings_add(ct, pgtbl, BOOT_CAPTBL_BOOTVM_PTE, "booter VM", mem_bootc_start(),
+	ret = boot_pgtbl_mappings_add(ct, BOOT_CAPTBL_SELF_PT, BOOT_CAPTBL_BOOTVM_PTE, "booter VM", mem_bootc_start(),
```
Comment (gparmer) [inline src/platform/i386/boot_comp.c]: I'm missing why this change has been made.  Why?

Is this so that both pgtbls are mapped into the booter pt?

File: src/platform/i386/boot_comp.c [hunk 9]
```diff
@@ -155,8 +158,10 @@ kern_boot_comp(void)
 	 * Need to account for the pages that will be allocated as
 	 * PTEs
 	 */
+	local_pgd = (pgtbl_t)chal_va2pa(mem_boot_alloc(1));
+	if (pgtbl_activate(ct, BOOT_CAPTBL_SELF_CT, BOOT_CAPTBL_SELF_LOCAL_PT, local_pgd, 0)) assert(0);
 	nkmemptes = boot_nptes(mem_utmem_end() - mem_boot_end());
-	ret = boot_pgtbl_mappings_add(ct, pgtbl, BOOT_CAPTBL_KM_PTE, "untyped memory", mem_boot_nalloc_end(nkmemptes),
+	ret = boot_pgtbl_mappings_add(ct, BOOT_CAPTBL_SELF_LOCAL_PT, BOOT_CAPTBL_KM_PTE, "untyped memory", mem_boot_nalloc_end(nkmemptes),
```
Comment (gparmer) [inline src/platform/i386/boot_comp.c]: Are you parameterizing which PT to use so that you can use it in the future for the global memory?

File: src/components/implementation/tests/micro_booter/micro_booter.c (hunk 1 above)
Comment (ryuxin) [inline src/components/implementation/tests/micro_booter/micro_booter.c]: Because in the future, I want to two meminfo, local_mi and global_mi. I think this explicit distinction makes the code more clear.
In addition,  it seems global memory will become hot research topic and we may heavily use global memory in the near future. I am not sure local one is still typical or not.
But if you want to keep the name unchanged, it is totally fine.

File: src/components/include/cos_kernel_api.h (hunk 2 above)
Comment (ryuxin) [inline src/components/include/cos_kernel_api.h]: no, "local" just refers to local memory.

File: src/kernel/include/shared/cos_types.h (hunk 3 above)
Comment (ryuxin) [inline src/kernel/include/shared/cos_types.h]: We definitely discussed this before. But before we put multiple page tables in the cos_info, which makes the code a little tedious, so want to simplify the code.
But now we have to use multiple mem_info. So again, I favor explicit separation between pgtbls for address translation and resource tracking.
And using multiple page tables consumes more memory, I do not think this is the issue we really care about now.
But if we use SELF_PT both for resource tracking and address translation, this will stress the address space resource of the manager.  Think about memory manager, if it managers 1G untyped memory, it just wastes lots of PTEs for the untyped memory, which limits the size of address space the manager can use. I probably miss something here.

File: src/components/lib/cos_kernel_api.c (hunk 4 above)
Comment (ryuxin) [inline src/components/lib/cos_kernel_api.c]: Yes,  @phani can help us to make sure when we use ci or meta.
I think a general rule is whenever you provide a memory capability (kmem here), you need to provide the corresponding page table who holds that capability.
I think in current code, all memory are allocated from the metta's page table.

File: src/platform/i386/boot_comp.c (hunk 9 above)
Comment (ryuxin) [inline src/platform/i386/boot_comp.c]: Yes, this replies to all your above related comments.
I want to parameterize to which pgtbl we add mapping.
So I change the argument to use the pgtbl capability instead of physical address of the pgtbl root.

**Comment by gparmer** [comment]
I'm glad that you're doing this iteratively.  This PR is mainly here to 1. put the pgtbl in the mem_info struct, and 2. use that appropriately when allocating resource.  The latter fixes where resources are allocated for a few kernel objects.

Is this accurate?  Is there more?

Your summary says that you're trying to provide multiple page tables, but I don't see that, so I'm not sure if that is part of this PR.  Is that true?

File: src/platform/i386/boot_comp.c (hunk 7 above)
Comment (ryuxin) [inline src/platform/i386/boot_comp.c]: No, this is used for the new page table we created to track local system memory.
See code L161 in the same file.
But if we decide to only use the "SELF_PT", I can remove it.

File: src/components/implementation/tests/micro_booter/micro_booter.c (hunk 1 above)
Comment (gparmer) [inline src/components/implementation/tests/micro_booter/micro_booter.c]: 1. that means that there are ton of modifications in this PR that make it very hard to go through for almost no gain.

2. That would make your improvements very limited in their intended use.  What if there are other devices that expose physical memory but don't expose global memory?  Xeon Phis are an example.

You're providing a general facility for using multiple physical page-tables for allocation.  Why bind it one one use-case?

File: src/components/include/cos_kernel_api.h (hunk 2 above)
Comment (gparmer) [inline src/components/include/cos_kernel_api.h]: See my previous comment.  I see no reason to broadly change all of this code for little reason.

File: src/components/lib/cos_kernel_api.c (hunk 4 above)
Comment (gparmer) [inline src/components/lib/cos_kernel_api.c]: I agree.  I've asked @phani to look at this to make sure.

File: src/kernel/include/shared/cos_types.h (hunk 3 above)
Comment (gparmer) [inline src/kernel/include/shared/cos_types.h]: I totally missed the point of this!  Do I have this right:  You're trying to put all the untyped memory into a *separate* page table instead of in the booter's PT?  If that's the case, then you're going in exactly the right direction.  We want that anyway.  If that's the case, we might as well start mapping at address 0.  Is this understanding correct?

**Comment by ryuxin** [comment]
My summary is perhaps inaccurate.
First, we try to find a approach who has the ability to support multiple page tables. Current approach is to extend mem_info to include a pgtbl_cap. If current code works, adding more page tables is trivial, as we can add more mem_info structs.
Second it uses mulitple page tables in this PR. As I have "SELF_PT"and "SELF_LOCAL_PT"

File: src/platform/i386/boot_comp.c (hunk 9 above)
Comment (gparmer) [inline src/platform/i386/boot_comp.c]: OK, I think i get the point of all of this.

Aside from the obvious issues with the commit (types, unused local variables, etc...), I think this is OK.  I don't like using `local` in code that is supposed to be generic (i.e. not only used for *your* functionality).  In the enum of capabilities, perhaps we can use `BOOT_CAPTBL_SELF_UNTYPED_PT`?

File: src/kernel/include/shared/cos_types.h (hunk 3 above)
Comment (ryuxin) [inline src/kernel/include/shared/cos_types.h]: Yes, I am exactly trying to put all the untyped memory into a separate page table (SELF_LOCAL_PT) instead of in the booter's SELF_PT.
But I did not start at address 0. I think keep this address unchanged makes it easy to debug. 
But it is necessary to change it to address 0.
After we agree on other parts, I will change it.

Comment (ryuxin) [inline src/kernel/include/shared/cos_types.h]: And keeping the mem_info variable name unchanged maybe better.
I think the key question we have is "do we want separate SELF_PT and SELF_LOCAL_PT"

**Comment by gparmer** [comment]
Yeah, I understand where you're going with all of this now.  As I mentioned i the previous comment, `BOOT_CAPTBL_SELF_UNTYPED_PT` is perhaps more appropriate instead of `LOCAL`

File: src/platform/i386/boot_comp.c (hunk 5 above)
Comment (ryuxin) [inline src/platform/i386/boot_comp.c]: I change the argument from pgtbl physical address to pgtbl cap
Does it make sense to you now?

File: src/kernel/include/shared/cos_types.h (hunk 3 above)
Comment (gparmer) [inline src/kernel/include/shared/cos_types.h]: yes.  see my other comments.

**Comment by ryuxin** [comment]
Ok, now we have those agreements.
1. keep the name of mem_info variable (mi) unchanged. This avoid a lot of code modification.
2. Use BOOT_CAPTBL_SELF_UNTYPED_PT. I think this is better. But it is still refer to a different page table from BOOT_CAPTBL_SELF_PT. The mapping in BOOT_CAPTBL_SELF_UNTYPED_PT starts at address 0.
The pending question is related to when we use "ci" and "meta".

File: src/platform/i386/boot_comp.c (hunk 5 above)
Comment (gparmer) [inline src/platform/i386/boot_comp.c]: I don't understand what this means.  I see no physical address.  I only see a type that was made ambiguous.  I don't understand your statement as it seems trivially wrong.

**Comment by gparmer** [comment]
I would appreciate if you could read your summary again and ask yourself how you could help us get to this level of understanding.  I don't have 2 hours for every pull request.  Especially one with 20 lines of changes.  It is very hard by looking at the code to really understand how it all fits together, and your current summary wasn't specific enough to really give me enough to grasp on to.

The other alternative is that I spend the 15 minutes I should, and you do another PR that I throw back at you, and we repeat N times.  That sounds much more frustrating for both of us.  Consider thinking about providing as much context as possible in the summary.  Organizing your summary will go along way toward making *other people* understand it.  Try saying "if the person reading this doesn't know anything about what this change is, how can I lead them into an understanding".

File: src/platform/i386/boot_comp.c (hunk 5 above)
Comment (ryuxin) [inline src/platform/i386/boot_comp.c]: the type "pgtbl_t" is defined as `typedef struct pgtbl * pgtbl_t;`
More specific, we can see the code `pgtbl_t pgtbl = (pgtbl_t)chal_va2pa(&boot_comp_pgd);`
So "pgtbl_t" is the physical address of the page table.
I change the argument to use capability instead of physical address.
We extensively use "capid_t" everywhere in the kernel to refer to a capability.
Why do you say it is ambiguous?

Comment (gparmer) [inline src/platform/i386/boot_comp.c]: I see now.  Thanks for explaining.  This all makes sense now.

File: src/kernel/include/shared/cos_types.h (hunk 3 above)
Comment (phanikishoreg) [inline src/kernel/include/shared/cos_types.h]: I'm missing the distinction of SELF_PT and UNTYPED_PT here, especially in case of Booter component. Isn't `pgtbl` of a component supposed to have all the pages that it has access to? How is separating it between two PT caps achieve this? Also, every time booter component or any component that has access to UNTYPED_PT, would then alloc, which will have to cpy from its UNTYPED_PT to SELF_PT to mean that it has access to its resources. 

One thing this simplifies I assume is, when we have a separate memmgr, we can just copy UNTYPED_PT cap to memmgr as opposed to copying as many pgtbl caps that it can manage resources for components, but that should not be much of a overhead if that happens only at boot-time. What am I missing here?

Comment (phanikishoreg) [inline src/kernel/include/shared/cos_types.h]: Also, having a separate comp_info for global memory would be much more useful because then, user can call the kernel api with different comp_info appropriately? If I need it to allocate memory in global memory, I would do `cos_page_bump_alloc(global_compinfo);` perhaps. But in the above design, wouldn't there be significant cos_kernel_api changes ?

File: src/components/lib/cos_kernel_api.c [hunk 10]
```diff
@@ -109,7 +110,7 @@ __mem_bump_alloc(struct cos_compinfo *__ci, int km)
 		/* are we dealing with a kernel memory allocation? */
 		syscall_op_t op = km ? CAPTBL_OP_MEM_RETYPE2KERN : CAPTBL_OP_MEM_RETYPE2USER;
 
-		if (call_cap_op(ci->pgtbl_cap, op, ret, 0, 0, 0)) return 0;
+		if (call_cap_op(ci->local_mi.pgtbl_cap, op, ret, 0, 0, 0)) return 0;
```
Comment (phanikishoreg) [inline src/components/lib/cos_kernel_api.c]: Once we do RETYPE this, this component has access to that page, should that mean it has to be copied to its `SELF_PT`? I'm confused what remains in `SELF_PT` otherwise.

File: src/components/lib/cos_kernel_api.c [hunk 11]
```diff
@@ -200,7 +201,7 @@ __capid_captbl_check_expand(struct cos_compinfo *ci)
 
 	printd("__capid_captbl_check_expand->pre-captblactivate (%d)\n", CAPTBL_OP_CAPTBLACTIVATE);
 	/* captbl internal node allocated with the resource provider's captbls */
-	if (call_cap_op(meta->captbl_cap, CAPTBL_OP_CAPTBLACTIVATE, captblcap, meta->pgtbl_cap, kmem, 1)) {
+	if (call_cap_op(meta->captbl_cap, CAPTBL_OP_CAPTBLACTIVATE, captblcap, meta->local_mi.pgtbl_cap, kmem, 1)) {
```
Comment (phanikishoreg) [inline src/components/lib/cos_kernel_api.c]: I guess, like Gabe mentioned before, `OP_CONS` which is called next must take in an additional parameter that indicates which `pgtbl/captbl` has this `captblcap`.

File: src/components/lib/cos_kernel_api.c [hunk 12]
```diff
@@ -298,7 +299,7 @@ __page_bump_valloc(struct cos_compinfo *ci)
 
 		/* PTE */
 		if (call_cap_op(meta->captbl_cap, CAPTBL_OP_PGTBLACTIVATE,
-				pte_cap, meta->pgtbl_cap, ptemem_cap, 1)) {
+				pte_cap, meta->local_mi.pgtbl_cap, ptemem_cap, 1)) {
```
Comment (phanikishoreg) [inline src/components/lib/cos_kernel_api.c]: `OP_CONS` must know `pte_cap` is from `meta`'s PTE, right? Like in capid expand.

File: src/components/lib/cos_kernel_api.c (hunk 4 above)
Comment (phanikishoreg) [inline src/components/lib/cos_kernel_api.c]: @gparmer it worked fine before because the threads were created from `vkernel` component and were copied to vm captbl, which as you said didn't use `meta`. And when the VMs started, they had their own untyped memory, so again the same no `meta` case there. 

@ryuxin this looks alright to me, but i've come to realize all the `ci->pgtbl_cap` are going to be replaced with `ci->local_mi.pgtbl_cap`, which pretty much removes `SELF_PT` from the picture. Isn't it?

File: src/components/lib/cos_kernel_api.c (hunk 10 above)
Comment (gparmer) [inline src/components/lib/cos_kernel_api.c]: I think it will only include the virtual memory of the booter.

File: src/components/lib/cos_kernel_api.c (hunk 11 above)
Comment (gparmer) [inline src/components/lib/cos_kernel_api.c]: still don't know if we actually want to do that.

File: src/components/lib/cos_kernel_api.c (hunk 4 above)
Comment (gparmer) [inline src/components/lib/cos_kernel_api.c]: I don't understand how you're making the second deduction, @phani.  SELF_PT has to be somewhere in all of the pgtbl_caps

Comment (phanikishoreg) [inline src/components/lib/cos_kernel_api.c]: I mean, all the `ci->pgtbl_cap` which is `SELF_PT` seem to get replaced with `local_mi.pgtbl_cap` (`UNTYPED_PT`), so I said that. 
But your last comment about `SELF_PT` having only virtual memory of the booter, I'm trying to understand that. In that case booter and/or memmgr, should have this distinction of `SELF_PT` and `UNTYPED_PT` and all other components only refer to `SELF_PT`? I'll look at the source code again to see how all this fits.

File: src/kernel/include/shared/cos_types.h (hunk 3 above)
Comment (ryuxin) [inline src/kernel/include/shared/cos_types.h]: "Isn't `pgtbl` of a component supposed to have all the pages that it has access to?"
Here "access" is a little ambiguous.
Take memory manager for example, it manages 1G memory. This does not mean the manager will actually use that 1G memory. Broadly there are two kinds of memory:
memory is used by the manager for its code data. Those memory are present in the page able referenced by SELF_PT.
other memory that is managed by the manager, but not actually accessed by the manager.
those memory are present in SELF_UNTYPED_PT.

"user can call the kernel api with different comp_info appropriately?"
This is also a little ambiguous.
A client component invokes a manager component without providing comp_info.
The manager finds the comp_info of that client based on its component id or other similar identifier.
It is possible for manager to maintain separate comp_info for each client. But I think comp_info is heavy weight in the case, as only we care about memory resource. comp_info also includes information about cap tables.

File: src/components/lib/cos_kernel_api.c (hunk 4 above)
Comment (ryuxin) [inline src/components/lib/cos_kernel_api.c]: `ci->pgtbl_cap`  is not `SELF_PT`
`meta->pgtbl_cap` is `SELF_PT`

File: src/components/lib/cos_kernel_api.c (hunk 10 above)
Comment (ryuxin) [inline src/components/lib/cos_kernel_api.c]: After RETYPE, we still can not access that page.
Only after we map the page into client's page table, client can access it. See CAPTBL_OP_MEMACTIVATE operation.

File: src/components/lib/cos_kernel_api.c (hunk 11 above)
Comment (ryuxin) [inline src/components/lib/cos_kernel_api.c]: I think the current semantics of OP_CONS is fine.
It connects a subtree in **our own** cap table to a node within anther cap table which needs to be explicated provided.
**our own** cap table is always implicitly provided, unless we want to modify it. 
Whenever we want to modify a cap table including our own cap table, we have to provide a cap to it.
But we read some caps in **our own** cap table, we do not need to provide a captbl_cap to **our own** cap table.

File: src/components/lib/cos_kernel_api.c (hunk 12 above)
Comment (ryuxin) [inline src/components/lib/cos_kernel_api.c]: See above comment. In this case, pte_cap has to be a cap in **our own** cap table.
We provide `meta->captbl_cap` because we modify **our own** cap table.
The interesting one is `ptemem_cap`. It is a cap in page table instead of cap table.
So we have to provide a cap to that page table, which is `meta->local_mi.pgtbl_cap`
Of course we need to be careful if you provide the correct page table

File: src/components/lib/cos_kernel_api.c (hunk 10 above)
Comment (phanikishoreg) [inline src/components/lib/cos_kernel_api.c]: @ryuxin you're right but if we name it `UNTYPED_PT`, I was thinking, once it's retyped, if it would still be in that PT or would be mapped into SELF_PT. I understand that it is mainly physical memory mappings that will be part of UNTYPED_PT, virtual mappings will be in SELF_PT.

File: src/components/lib/cos_kernel_api.c (hunk 11 above)
Comment (phanikishoreg) [inline src/components/lib/cos_kernel_api.c]: `captblcap` here is the new capability id that was created at the `frontier` not referring to our own captbl. That which was just activated at this step, will need to be `OP_CONS`ed into the `ci->captbl_cap` where we are trying to extend at `captblid_add`. I can see that there is no consistency check for this address in the kernel on `OP_CONS`, I wonder if that is required.

File: src/components/lib/cos_kernel_api.c (hunk 4 above)
Comment (phanikishoreg) [inline src/components/lib/cos_kernel_api.c]: `SELF_PT` is the offset in any captbl in my understanding. It wasn't literal meaning of "SELF" that I was referring to. So, `ci->pgtbl_cap` is `SELF_PT` of `ci`, same way for `meta`. 

But I understand now what you mean by `SELF_PT`. And I also understand that we are going to have physical memory mappings in `UNTYPED_PT` and virtual memory mappings in `SELF_PT`, which is what we mostly use `SELF_PT` for anyway. Physical memory pages are only alloced and mapped in `memsrc`, so it should not impact the API much, I believe.

File: src/kernel/include/shared/cos_types.h (hunk 3 above)
Comment (phanikishoreg) [inline src/kernel/include/shared/cos_types.h]: Well, I'm still trying to comprehend this whole idea of `UNTYPED_PT` distinction.
But I think the idea here is that, the `UNTYPED_PT` actually has mappings to the physical memory and `SELF_PT` has for virtual memory, which may as well be the memory that is used by memmgr or any component. Isn't it?

**Comment by gparmer** [comment]
Not sure if closing this was the right thing to do.  Now we can't have
record of the conversation, and I can't reply to Phani.  Oh well.
g

File: src/kernel/include/shared/cos_types.h (hunk 3 above)
Comment (ryuxin) [inline src/kernel/include/shared/cos_types.h]: Yes, I think so. SELF_PT tracks virtual memory of the manager

File: src/components/lib/cos_kernel_api.c (hunk 4 above)
Comment (ryuxin) [inline src/components/lib/cos_kernel_api.c]: More concrete, `SELF_PT` is a fixed value while `ci->pgtbl_cap` is a variable.
But you are right. both `ci->pgtbl_cap` in manager's cap table and `SELF_PT` in ci's cap table refers to the same page table
